### ISSUE TITLE
React Native fast refresh problems

### ISSUE THREAD

**Issue by thomasgosse** [pmndrs/zustand#210]
Hi !

I'm using Zustand v3 with React Native and everything works great. 
Troubles comes when I hit save and app is fast-refreshed: my store is emptied and no update is trigger anymore after that. The only way to make this work again is to reload the whole app.
This weird behavior **appears only when I hit save on my useStore file (which create store with create())**

I'm doing something like this:
```javascript
// App.js
const App = () => {
  const getLists = useStore((state) => state.getLists);
  const lists = useStore((state) => state.lists);
  useEffect(() => {
    getLists();
  }, []);
 
  console.log(lists);
  ...
}
```
the getLists method:
```javascript
  // useStore.js
  getLists: async () => {
      const lists = await getLists();
      set({ lists });
  },
```
the addList method:
```javascript
  // useStore.js
  addList: (list) => {
      const lists = [...get().lists];
      lists.push(list);
      set({ lists });
  },
```

So if I save my useStore.js, fast refresh comes in, and then if I try to add with my addList method, console.log is not called because the component is not rendered any more (but I think it should because I updated lists).
I confirm that this experiment works when I'm not using fast refresh of when I'm not saving useStore.js (console log is called so App is re-rendered).

Let me know if you need more information or a RN repo to reproduce this !

Thanks for this nice library btw 😃

**Comment by dai-shi** [comment]
Hey, thanks for reporting.
First question: Is this only with RN? You can't reproduce the issue with codesandbox (dom), can you?

**Comment by thomasgosse** [comment]
Unfortunately I can't reproduce it with codesandbox... I also use Zustand in a regular React webapp but I can't reproduce this. So I think it's only with RN.

**Comment by NorweskiDrwal** [comment]
@thomasgosse
That would be a RN debugger that you're using. I happened to add [react-native-debugger](https://github.com/jhen0409/react-native-debugger) to my Expo/Zustand project and everything went bonkers (disabling fast refresh doesn't help :D).
I noticed it after creating a theming service for my app, when I actually needed the state :D I went through all stages of grief, along with feeling like a shit dev, who can't even `React.useState()` properly before discovering it.
If I kill a RND it works fine, but dev-ex is worse ;) (but you can always console log the fuck of it, aside from network calls :D)

@dai-shi
Could the above be caused by Zustand's middleware/devtools implementation within the library itself?
I can confirm that `react-native-debugger` breaks the state with devtools added to the `create()` and without it.
It doesn't work with Expo's out-of-the-box debug either. So, a fair assumption is: if anything's connected to RN and running on `:19001` port, it'll break the state. Probably some underlaying issue, but my skills are weak and I don't know enough to elaborate, or come up with more probable conclusions

**Comment by dai-shi** [comment]
Thanks guys! I'm not sure if I understand the issue fully yet. Hope someone else to chime in to elaborate.

**Comment by NorweskiDrwal** [comment]
@dai-shi 
I hope the screen capture will make it a bit more clear. 
TL;DR: Zustand doesn't seem to work with RN debugging tools.

![expo-zustand](https://user-images.githubusercontent.com/30369902/96355102-4ddc4f00-10de-11eb-950f-bd34913e25a7.gif)

**Comment by dai-shi** [comment]
Thanks for the screen capture.

> Possibly because it lives outside of React and RN debuggers can't see it?

But, react-redux works in this scenario, doesn't it?

**Comment by NorweskiDrwal** [comment]
It does, that's why it's weird

**Comment by thomasgosse** [comment]
@NorweskiDrwal in my case RN Debugger works nicely ! But I'm not using Expo + RN, only RN, so maybe expo causes this ?

**Comment by thomasgosse** [comment]
@dai-shi Do you think a minimal React-Native repo with the behavior would help you ?

**Comment by dai-shi** [comment]
Certainly, it would. Honestly, I'm not so sure about the solution. My assumption is fast-refresh may not behave perfectly if you modify store file, but that should be the same with codesandbox. I mean I don't have a guess, currently. Reproduction should be helpful anyway for me and someone else. Anyone else coming in???

**Comment by thomasgosse** [comment]
Ok so while I was trying to create a demo repo, I think I isolated what was the problem in my app.

In order to achieve clear/dark modes, I created a custom hook:
```javascript
import { useEffect, useState } from 'react';
import { useColorScheme } from 'react-native';

const defaultColors =..., defaultVariables=...;

export default function useColors() {
  const colorScheme = useColorScheme();
  const [colors, setColors] = useState(defaultColors);
  const [variables] = useState(defaultVariables);
  useEffect(() => {
    if (colorScheme === 'dark') {
      const darkColors = {
         //.....
      };
      setColors(darkColors);
    } else {
      setColors(defaultColors);
    }
  }, [colorScheme]);

  return { colors, variables };
}
```
When I was doing some test from a new repo, everything was correct and I couldn't reproduce the "bug". 
So I tried to isolate some pieces of my own code. I tried to remove the useColorScheme() hook, which immediately made the "bug" disappear !
It is super weird...

At this point I'm not sure if the issue comes from Zustand or RN's useColorScheme hook !

**Comment by thomasgosse** [comment]
I create a repo with the weird behavior:
https://github.com/thomasgosse/RNFastRefreshZustand

You can try it both on iOS and Android.
I put instructions on src/components/hooks/useTheme.js
Keep in mind that this happen only when you hit save on the src/components/store/useStore.js file !

**Comment by dai-shi** [comment]
Thanks for the repro. Can confirm the behavior in iOS simulator.
It's an interesting behavior. I first thought useEffect doesn't fire, but it does.
There might be some unknown fast refresh behavior, but there doesn't seem to be anything we can do at the moment. I wonder if we can repro the behavior without useColorScheme.

**Comment by thomasgosse** [comment]
I don't know either ! The thing is that once you have fast refreshed and the behavior happened, the state won't update afterwards.
Anyway, I used Appearance from react-native and create my own "useColorScheme" so for my problem is solved for now.
We can leave this open, and see if other people notice the same behavior, or close it as you want 😀 (in both case this issue is known).

**Comment by dai-shi** [comment]
Yeah, let's leave it open for a while. Thanks for your reporting anyway!

**Comment by worstpractice** [comment]
I have to chime in by just mentioning that zustand and fast refresh has worked flawlessly for me in my current Expo project.

I'm using the managed Expo workflow, not bare or ejected.

It's a greenfield project, and daily development has been ongoing for a little over two months now.

[There are some pitfalls to be aware of when using fast refresh](https://reactnative.dev/docs/fast-refresh), but they can be avoided, as I have.

For example, exporting anything but one (1) functional component from a single file drastically changes how fast refresh behaves. As do the mere existence of class components, and so on. Fast refresh can be quite a different experience depending on how you import and export things.

I also happen to use the react native debugger, and have not run into any issues with it. Though I must be clear: I have barely had to use the debugger thus far, simply because no bug has been evasive enough to demand I use the debugger (fingers crossed that this state of affairs can last).

I'm developing exclusively on Android physical devices for now. I take care to never have the debugging mode enabled without having an extremely good and specific reason, because just having remote debugging enabled in the device basically (for all intents and purposes) changes how everything works in React Native. It is also much slower.

One piece of advice I always have for any React Native project is to clear the transform cache at every "yarn start" (or similar). By default, incorrect state survives CTRL+C:ing the bundler and restarting it.

I avoid bugs every day by having a start script that reads:
json```
"start": "expo start --clear"
```

BrowserStack also maintains a list of all the things in React Native that are completely different in the iOS simulator, versus a physical device. [Do not trust the simulator for anything critical, I must warn.](https://www.browserstack.com/test-on-ios-simulator)

I am using zustand 3.2.0, expo 40.0.0 and react 16.13.1.

**Comment by worstpractice** [comment]
@dai-shi I should make sure to point out the existence of the `// @refresh reset` pragma.

Unless babel handles it, I believe it to be a [Metro](https://facebook.github.io/metro/) pragma — though I imagine any webpack-based Metro replacement (e.g. [Haul](https://github.com/callstack/haul)) would support it as well.

From the docs:
> Sometimes you might want to _force_ the state to be reset, and a component to be remounted. For example, this can be handy if you're tweaking an animation that only happens on mount. To do this, you can add `// @refresh reset` anywhere in the file you're editing. This directive is local to the file, and instructs Fast Refresh to remount components defined in that file on every edit.

Assuming this pragma will safely affect only react native projects, I wonder if the issue could be solved by inserting it somewhere.

Note that the pragma **does not** affect normal react (re)rendering or component mounting. AFAICT the pragma exclusively interacts with fast refresh, which naturally does not exist in production.

If I have the time it would be interesting to sprinkle the pragma inside zustand and observe what happens to fast refresh, though again I have had no issue so would likely not observe anything useful.

**Comment by dai-shi** [comment]
@worstpractice 

https://reactnative.dev/docs/fast-refresh

It's very good to know. As I understand, `@refresh react` is not only for RN. I wasn't able to find a document though.

BTW, the issue in zustand is not force refreshing. It's the other way around.

**Comment by dai-shi** [comment]
Let me close this issue as it's a new year, and it seems Fast Refresh basically works fine in RN.
